test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+""" Test node eviction logic
+
+When the number of peers has reached the limit of maximum connections,
+the next connecting inbound peer will trigger the eviction mechanism.
+We cannot currently test the parts of the eviction logic that are based on
+address/netgroup since in the current framework, all peers are connecting from
+the same local address. See Issue #14210 for more info.
+
+Therefore, this test is limited to the remaining protection criteria:
+1.) Protect 4 nodes having sent us a block most recently
+2.) Protect 4 nodes having sent us a transaction most recently
+3.) Protect 8 nodes with the smallest minimum ping
+"""
+
+import time
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.mininode import P2PInterface, P2PDataStore
+from test_framework.util import assert_equal, wait_until
+from test_framework.blocktools import create_block, create_coinbase
+from test_framework.messages import CTransaction, FromHex, msg_pong, msg_tx
+
+
+class SlowP2PDataStore(P2PDataStore):
+    def on_ping(self, message):
+        time.sleep(0.1)
+        self.send_message(msg_pong(message.nonce))
+
+class SlowP2PInterface(P2PInterface):
+    def on_ping(self, message):
+        time.sleep(0.1)
+        self.send_message(msg_pong(message.nonce))
+
+
+class P2PEvict(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        # The choice of 32 maxconnections results in a maximum of 21 inbound connections
+        # (32 - 10 outbound - 1 feeler). 20 inbounds peers are protected from eviction.
+        self.extra_args = [['-maxconnections=32']]
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def run_test(self):
+        self.protected_peers = set() # peers that we expect to be protected from eviction
+        current_peer = 0
+        node = self.nodes[0]
+        node.generatetoaddress(101, node.get_deterministic_priv_key().address)
+
+        # The first 4 peers send us a block, protecting them from eviction.
```

Comment:
These single line comments above each block would be more helpful as logs. Rationale https://github.com/bitcoin/bitcoin/blob/master/test/functional/README.md#general-test-writing-advice
```suggestion
        self.log.info("Create the first 4 peers and protect them from eviction by sending us a block.")
```

Reply:
Done, thanks!